arbnode/message_pruner.go

```diff
@@ -25,107 +28,206 @@ type MessagePruner struct {
-		staker:              staker,
 		config:              config,
 	}
 }
 
 func (m *MessagePruner) Start(ctxIn context.Context) {
 	m.StopWaiter.Start(ctxIn, m)
-	m.CallIteratively(m.prune)
 }
 
-func (m *MessagePruner) prune(ctx context.Context) time.Duration {
-	latestConfirmedNode, err := m.staker.Rollup().LatestConfirmed(
-		&bind.CallOpts{
-			Context:     ctx,
-			BlockNumber: big.NewInt(int64(rpc.FinalizedBlockNumber)),
-		})
+func (m *MessagePruner) UpdateLatestStaked(count arbutil.MessageIndex, globalState validator.GoGlobalState) {
+	locked := m.pruningLock.TryLock()
+	if !locked {
+		return
+	}
+
+	if m.lastPruneDone.Add(m.config().MessagePruneInterval).After(time.Now()) {
+		m.pruningLock.Unlock()
+		return
+	}
+	err := m.LaunchThreadSafe(func(ctx context.Context) {
+		defer m.pruningLock.Unlock()
+		err := m.prune(ctx, count, globalState)
+		if err != nil {
+			log.Error("error while pruning", "err", err)
+		}
+	})
 	if err != nil {
-		log.Error("error getting latest confirmed node", "err", err)
-		return m.config().MessagePruneInterval
+		log.Info("failed launching prune thread", "err", err)
+		m.pruningLock.Unlock()
 	}
-	nodeInfo, err := m.staker.Rollup().LookupNode(ctx, latestConfirmedNode)
+}
+
+// looks for batch posting report starting from delayed message delayedMsgStart
+// returns number of batch for which report was found (meaning - it should not be pruned)
+// if not found - returns maxUint64 (no limit on pruning)
+func (m *MessagePruner) findBatchReport(ctx context.Context, delayedMsgStart uint64) (uint64, error) {
+	searchLimit := m.config().SearchBatchReportLimit
+	if searchLimit < 0 {
+		return math.MaxUint64, nil
+	}
+	delayedCount, err := m.inboxTracker.GetDelayedCount()
 	if err != nil {
-		log.Error("error getting latest confirmed node info", "node", latestConfirmedNode, "err", err)
-		return m.config().MessagePruneInterval
+		return 0, err
+	}
+	if delayedCount <= delayedMsgStart {
+		return 0, errors.New("delayedCount behind pruning target")
+	}
+	searchUpTil := delayedCount
+	searchUpLimit := delayedMsgStart + uint64(searchLimit)
+	if searchLimit > 0 && searchUpLimit < searchUpTil {
+		searchUpTil = searchUpLimit
 	}
-	endBatchCount := nodeInfo.Assertion.AfterState.GlobalState.Batch
-	if endBatchCount == 0 {
-		return m.config().MessagePruneInterval
+	for delayed := delayedMsgStart; delayed < searchUpTil; delayed++ {
+		if ctx.Err() != nil {
+			return 0, ctx.Err()
+		}
+		msg, err := m.inboxTracker.GetDelayedMessage(delayed)
+		if err != nil {
+			return 0, err
+		}
+		if msg.Header.Kind == arbostypes.L1MessageType_BatchPostingReport {
+			_, _, _, batchNum, _, _, err := arbostypes.ParseBatchPostingReportMessageFields(bytes.NewReader(msg.L2msg))
+			if err != nil {
+				return 0, fmt.Errorf("trying to parse batch-posting report: %w", err)
+			}
+			return batchNum, nil
+		}
 	}
-	endBatchMetadata, err := m.inboxTracker.GetBatchMetadata(endBatchCount - 1)
+	searchDownLimit := uint64(0)
```

Comment:
I don't think this should be reachable unless the searchLimit is hit. The inbox reader reads delayed messages and batches at the same time (and in fact we insert delayed messages into the inbox tracker before batches), so if we have a batch, we should have its corresponding batch posting report because that's generated in the same L1 block. It would simplify the code to just error here instead of doing the downwards search, and I think that'd be safe.